File: InputTestCode/Swift/SwiftTour/6_ProtocolsExtensions.swift
// Protocols
protocol ExampleProtocol {
    var simpleDescription: String { get }
    mutating func adjust()
}

class SimpleClass: ExampleProtocol {
    var simpleDescription: String = "A very simple class."
    var anotherProperty: Int = 69105
    func adjust() {
        simpleDescription += "  Now 100% adjusted."
    }
}

func generateSimpleClass() -> String {
    let a = SimpleClass()
    a.adjust()

    let protocolValue: ExampleProtocol = a

    return a.simpleDescription
}

struct SimpleStructure: ExampleProtocol {
    var simpleDescription: String = "A simple structure"
    mutating func adjust() {
        simpleDescription += " (adjusted)"
    }
}

func generateSimpleStruct() -> String {
    var b = SimpleStructure()
    b.adjust()

    return b.simpleDescription
}

// Extensions
extension Int: ExampleProtocol {
    var simpleDescription: String {
        return "The number \(self)"
    }
    mutating func adjust() {
        self += 42
    }
}

let seven = 7
let sevenDescription = seven.simpleDescription
var eight = 8

func makeUseOfExtension() {
    eight.adjust()
}

File: InputTestCode/Kotlin/SwiftTour/6_ProtocolsExtensions.kt
// import java.lang.Error

// Protocols
interface ExampleProtocol {
    var simpleDescription: String

    open fun adjust()
}

class SimpleClass: ExampleProtocol {
    override var simpleDescription: String = "A very simple class."
    var anotherProperty: Int = 69105

    override fun adjust() {
        simpleDescription += "  Now 100% adjusted."
    }
}

fun generateSimpleClass(): String {
    val a = SimpleClass()
    a.adjust()

    val protocolValue: ExampleProtocol = a

    return a.simpleDescription
}

data class SimpleStructure(val value1: Any?): ExampleProtocol {
    override var simpleDescription: String = "A simple structure"
    override fun adjust() {
        simpleDescription += " (adjusted)"
    }
}

fun generateSimpleStruct(): String {
    val b = SimpleStructure(null)
    b.adjust()
    return b.simpleDescription
}

// Extensions
val Int.simpleDescription: String
    get() = "The number ${this}"
// fun Int.adjust(): Int = this + 42

val seven = 7
val sevenDescription = seven.simpleDescription
var eight = 8

// fun makeUseOfExtension() {
//     eight = eight.adjust()
// }